engine/src/main/java/io/camunda/zeebe/streamprocessor/DirectProcessingResultBuilder.java

```diff
@@ -54,6 +68,21 @@ public ProcessingResultBuilder appendRecord(
       final RejectionType rejectionType,
       final String rejectionReason,
       final RecordValue value) {
+
+    final ValueType valueType = typeRegistry.get(value.getClass());
+    if (valueType == null) {
+      // usually happens when the record is not registered at the TypedStreamEnvironment
+      throw new IllegalStateException("Missing value type mapping for record: " + value.getClass());
+    }
+
+    if (value instanceof UnifiedRecordValue unifiedRecordValue) {
+      mutableRecordBatch.appendRecord(
+          key, -1, type, intent, rejectionType, rejectionReason, valueType, unifiedRecordValue);
+    } else {
+      throw new IllegalStateException(
+          String.format("The record value %s is not a UnifiedRecordValue", value));
+    }
+
     streamWriter.appendRecord(key, type, intent, rejectionType, rejectionReason, value);
```

Comment:
No, the streamWriter is reseted in the ProcessingStateMachine. The problem here is I can't yet remove it because I have to migrate the TaskResult first to the RecordBatch, then I can remove that part.